File: Chatmvp/app/src/main/java/nosql/anew/chatmvp/utils/FirebaseManager.java
package nosql.anew.chatmvp.utils;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.HashMap;
import java.util.Map;

import nosql.anew.chatmvp.interfaces.FirebaseCallBacks;

/**
 * Created by saksham on 26/6/17.
 */

public class FirebaseManager implements ChildEventListener{
    private volatile static FirebaseManager sFirebaseManager;
    private DatabaseReference mMessageReference;
    private FirebaseCallBacks mCallbacks;

    public static synchronized FirebaseManager getInstance(String roomName, FirebaseCallBacks callBacks) {
        if(sFirebaseManager == null) {
            synchronized (FirebaseManager.class) {
                sFirebaseManager = new FirebaseManager(roomName,callBacks);
            }
        }
        return sFirebaseManager;
    }

    private FirebaseManager(String roomName, FirebaseCallBacks callBacks){
        mMessageReference = FirebaseDatabase.getInstance().getReference().child(roomName);
        this.mCallbacks =callBacks;
    }

    public void addMessageListeners(){
        mMessageReference.addChildEventListener(this);
    }

    public void removeListeners(){
        mMessageReference.removeEventListener(this);
    }

    @Override
    public void onChildAdded(DataSnapshot dataSnapshot, String s) {
        mCallbacks.onNewMessage(dataSnapshot);
    }

    @Override
    public void onChildChanged(DataSnapshot dataSnapshot, String s) {

    }

    @Override
    public void onChildRemoved(DataSnapshot dataSnapshot) {

    }

    @Override
    public void onChildMoved(DataSnapshot dataSnapshot, String s) {

    }

    @Override
    public void onCancelled(DatabaseError databaseError) {

    }

    public void sendMessageToFirebase(String message) {
        Map<String,Object> map=new HashMap<>();
        map.put("text",message);
        map.put("time",System.currentTimeMillis());
        map.put("senderId", FirebaseAuth.getInstance().getCurrentUser().getUid());

        String keyToPush= mMessageReference.push().getKey();
        mMessageReference.child(keyToPush).setValue(map);
    }

    public void destroy() {
        sFirebaseManager=null;
        mCallbacks =null;
    }
}

File: Chat-Kotlinmvc/app/src/main/java/nosql/anew/kotlinmvc/managers/FirebaseManager.kt
package nosql.anew.kotlinmvc.managers

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import nosql.anew.kotlinmvc.interfaces.FirebaseCallBacks
import java.util.HashMap

/**
 * Created by Sam on 12/04/18.
 */
class FirebaseManager private constructor(roomName: String, private var mCallbacks: FirebaseCallBacks?) : ChildEventListener {

    private val mMessageReference: DatabaseReference

    init {
        mMessageReference = FirebaseDatabase.getInstance().reference.child(roomName)
    }

    fun addMessageListeners() {
        mMessageReference.addChildEventListener(this)
    }

    fun removeListeners() {
        mMessageReference.removeEventListener(this)
    }

    override fun onChildAdded(dataSnapshot: DataSnapshot?, s: String?) {
        mCallbacks!!.onNewMessage(dataSnapshot!!)
    }

    override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {

    }

    override fun onChildRemoved(dataSnapshot: DataSnapshot) {

    }

    override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {

    }

    override fun onCancelled(databaseError: DatabaseError) {

    }

    fun sendMessageToFirebase(message: String) {
        val map = HashMap<String, Any>()
        map["text"] = message
        map["time"] = System.currentTimeMillis()
        map["senderId"] = FirebaseAuth.getInstance().currentUser!!.uid

        val keyToPush = mMessageReference.push().key
        mMessageReference.child(keyToPush).setValue(map)
    }

    fun destroy() {
        sFirebaseManager = null
        mCallbacks = null
    }

    companion object {
        @Volatile
        private var sFirebaseManager: FirebaseManager? = null

        @Synchronized
        fun getInstance(roomName: String, callBacks: FirebaseCallBacks): FirebaseManager? {
            if (sFirebaseManager == null) {
                synchronized(FirebaseManager::class.java) {
                    sFirebaseManager = FirebaseManager(roomName, callBacks)
                }
            }
            return sFirebaseManager
        }
    }
}
